clash-lib/src/Clash/Core/HasType.hs

```diff
@@ -0,0 +1,245 @@
+{-|
+
+Utility class to extract type information from data which has a type.
+-}
+
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+module Clash.Core.HasType
+  ( HasType(..)
+  , InferType(..)
+  , applyTypeToArgs
+  , piResultTy
+  , piResultTyMaybe
+  , piResultTys
+  ) where
+
+import qualified Data.Text as Text (isInfixOf)
+import Data.Text.Prettyprint.Doc (line)
+import GHC.Stack (HasCallStack)
+
+import Clash.Core.DataCon (DataCon(dcType))
+import Clash.Core.FreeVars
+import Clash.Core.Literal (Literal(..))
+import Clash.Core.Name (Name(nameOcc))
+import Clash.Core.Pretty
+import Clash.Core.Subst
+import Clash.Core.Term (Term(..), IsMultiPrim(..), PrimInfo(..), collectArgs)
+import Clash.Core.TyCon (TyConMap)
+import Clash.Core.Type
+import Clash.Core.TysPrim
+import Clash.Core.Var (Var(varType))
+import Clash.Core.VarEnv
+import Clash.Debug (debugIsOn)
+import Clash.Util (pprPanic)
+import qualified Clash.Util.Interpolate as I
+
+class HasType a where
+  coreTypeOf :: a -> Type
+
+instance HasType DataCon where
+  coreTypeOf = dcType
+
+instance HasType Literal where
+  coreTypeOf = \case
+    IntegerLiteral _ -> integerPrimTy
+    IntLiteral _ -> intPrimTy
+    WordLiteral _ -> wordPrimTy
+    StringLiteral _ -> stringPrimTy
+    FloatLiteral _ -> floatPrimTy
+    DoubleLiteral _ -> doublePrimTy
+    CharLiteral _ -> charPrimTy
+    Int64Literal _ -> int64PrimTy
+    Word64Literal _ -> word64PrimTy
+    NaturalLiteral _ -> naturalPrimTy
+    ByteArrayLiteral _ -> byteArrayPrimTy
+
+instance HasType PrimInfo where
+  coreTypeOf pr =
+    case primMultiResult pr of
+      SingleResult -> primType pr
+      MultiResult
+        | let (primArgs, primResTy) = splitFunForallTy (primType pr)
+        , TyConApp tupTcNm tupArgs <- tyView primResTy
+        , Text.isInfixOf "GHC.Tuple.(" (nameOcc tupTcNm)
+        -> mkPolyFunTy primResTy (primArgs <> fmap Right tupArgs)
+
+        | otherwise
+        -> error "PrimInfo.coreTypeOf: MultiResult primitive without tuple type"
```

Comment:
This really needs a note here to explain why it behaves as it does, e.g. the rationale behind the old `multiPrimType` that this replaces

Reply:
Updated: the doc in `Clash.Normalize.Transformation.MultiPrim` was extended and this now points to that.